### PR TITLE
Preparation for the QUIC proxy peering implementation

### PR DESCRIPTION
**Pull request by espadolini**
This PR rearranges some of `lib/proxy/peer` and related packages to better support the addition of different transports for proxy peering. Prerequisite for #47587.

### CODE REVIEW

File: lib/proxy/peer/client.go [hunk 1]
```diff
@@ -686,8 +661,8 @@ func (c *Client) getConnections(proxyIDs []string) ([]clientConn, bool, error) {
 			continue
 		}
 
-		_, supportsQuic := proxy.GetLabel(types.ProxyPeerQUICLabel)
-		conn, err := c.connect(id, proxy.GetPeerAddr(), supportsQuic)
+		supportsQUIC, _ := proxy.GetLabel(types.UnstableProxyPeerQUICLabel)
+		conn, err := c.connect(id, proxy.GetPeerAddr(), supportsQUIC == "yes")
```
Comment (codingllama) [inline lib/proxy/peer/client.go:665]: Same about guarding the feature?

File: lib/proxy/peer/internal/clientconn.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+		src net.Addr,
+		dst net.Addr,
+		tunnelType types.TunnelType,
+	) (net.Conn, error)
+
+	// Close closes all connections and releases any background resources
+	// immediately.
+	Close() error
+
+	// Shutdown waits until all connections are closed or the context is done,
```
Comment (codingllama) [inline lib/proxy/peer/internal/clientconn.go:47]: Is it worth mentioning the behavior regarding new connections? Is it uniform/prescribed enough to be here?

File: lib/proxy/peer/internal/tls.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+// VerifyPeerCertificateIsProxy is a function usable as a
+// [tls.Config.VerifyPeerCertificate] callback to enforce that the connected TLS
+// client is using Proxy credentials.
+func VerifyPeerCertificateIsProxy(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+	if len(verifiedChains) < 1 {
+		return trace.AccessDenied("missing client certificate (this is a bug)")
+	}
+
+	clientCert := verifiedChains[0][0]
```
Comment (codingllama) [inline lib/proxy/peer/internal/tls.go:39]: Should we check `len(verifiedChains[0])` first, just to be safe?

File: lib/proxy/peer/internal/tls.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+// VerifyPeerCertificateIsSpecificProxy returns a function usable as a
+// [tls.Config.VerifyPeerCertificate] callback to enforce that the connected TLS
+// server is using Proxy credentials and has the expected host ID.
+func VerifyPeerCertificateIsSpecificProxy(peerID string) func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+	return func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+		if len(verifiedChains) < 1 {
+			return trace.AccessDenied("missing server certificate (this is a bug)")
+		}
+
+		clientCert := verifiedChains[0][0]
```
Comment (codingllama) [inline lib/proxy/peer/internal/tls.go:60]: Same for length check.

File: lib/proxy/peer/internal/tls.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+		}
+		return nil
+	}
+}
+
+const duplicatePeerMsg = "Detected multiple Proxy Peers with the same public address when connecting to a Proxy which can lead to inconsistent state and problems establishing sessions. For best results ensure that `peer_public_addr` is unique per proxy and not a load balancer."
+
+// LogDuplicatePeer should be used to log a message if a proxy peering client
+// connects to a Proxy that did not have the expected host ID.
+func LogDuplicatePeer(ctx context.Context, log *slog.Logger, level slog.Level, args ...any) {
```
Comment (codingllama) [inline lib/proxy/peer/internal/tls.go:81]: Suggestion: move to const to method (smallest scope needed).

```suggestion
// LogDuplicatePeer should be used to log a message if a proxy peering client
// connects to a Proxy that did not have the expected host ID.
func LogDuplicatePeer(ctx context.Context, log *slog.Logger, level slog.Level, args ...any) {
	const duplicatePeerMsg = "" +
		"Detected multiple Proxy Peers with the same public address when connecting to a Proxy which can lead to inconsistent state and problems establishing sessions. " +
		"For best results ensure that `peer_public_addr` is unique per proxy and not a load balancer."
```

File: lib/config/configuration.go [hunk 6]
```diff
@@ -2727,6 +2727,11 @@ func Configure(clf *CommandLineFlags, cfg *servicecfg.Config, legacyAppFlags boo
 		cfg.DebugService.Enabled = false
 	}
 
+	// TODO(espadolini): allow this when the implementation is merged
+	if false && os.Getenv("TELEPORT_UNSTABLE_QUIC_PROXY_PEERING") == "yes" {
```
Comment (codingllama) [inline lib/config/configuration.go:2731]: WDYT about having a QUICEnabled feature flag instead of the "false" here?

**Comment by codingllama** [comment]
Thanks for the small/focused refactors!

File: lib/proxy/peer/client.go [hunk 7]
```diff
@@ -485,8 +460,8 @@ func (c *Client) updateConnections(proxies []types.Server) error {
 		}
 
 		// establish new connections
-		_, supportsQuic := proxy.GetLabel(types.ProxyPeerQUICLabel)
-		conn, err := c.connect(id, proxy.GetPeerAddr(), supportsQuic)
+		supportsQUIC, _ := proxy.GetLabel(types.UnstableProxyPeerQUICLabel)
+		conn, err := c.connect(id, proxy.GetPeerAddr(), supportsQUIC == "yes")
```
[the remark this answers is not in the document]
Comment (espadolini) [reply lib/proxy/peer/client.go:464]: I don't think so - `supportsQUIC` refers to what the peer proxy advertises in its heartbeat here, not what the current instance will decide what to do about it; `connect` will also check for local availability of a QUIC transport before using QUIC, falling back to gRPC even if the peer supports QUIC.

File: lib/proxy/peer/internal/tls.go (hunk 3 above)
Comment (espadolini) [reply lib/proxy/peer/internal/tls.go:39]: `verifiedChains` is a slice of nonempty slices, as per the `crypto/tls` docs. I think I'd rather segfault than live in a world where my TLS library is violating an invariant around certificate chain verification - wdyt?

File: lib/proxy/peer/internal/clientconn.go (hunk 2 above)
Comment (espadolini) [reply lib/proxy/peer/internal/clientconn.go:47]: It's not uniform enough and it also doesn't currently matter, we call `Shutdown` on clientconns long after anything is or should be creating more connections. When (if?) we make shutdown more parallel and saner we can revisit it a bit.

Comment (codingllama) [reply lib/proxy/peer/internal/clientconn.go:47]: Thanks.

**Review by codingllama — APPROVED** [review]

**Review by rosstimothy — APPROVED** [review]

**Pull request merged**